### PR TITLE
GH-2080 Prevent local storage to maintain list of files which have no markers associated to it

### LINKED ISSUE #2080: [localstorage] Do not store files with no markers into the local storage.
We should prevent storing empty markers into the local storage. This will reduce the information found in the local storage.
Hint: Look at the saveMarkersToStorage()

### PR DESCRIPTION
**Pull request by lmcbout**
Fixes #2080 

Signed-off-by: Jacques Bouthillier <jacques.bouthillier@ericsson.com>

### CODE REVIEW

**Comment by marcdumais-work** [comment]
I wonder if it's useful for ```marker-manager``` to keep an empty problem-marker collection, for a given URI? For example, if a file has errors, that are then fixed, would there be a reason to keep the marker collection for that file's URI? Or could it just then be cleared? 

Clearing marker collections for URIs that no longer have problem markers would be another way to address #2080

File: packages/markers/src/browser/marker-manager.ts [hunk 1]
```diff
@@ -148,18 +148,19 @@ export abstract class MarkerManager<D extends object> {
             for (const [uri, collection] of this.uri2MarkerCollection.entries()) {
                 const ownerEntries: Owner2MarkerEntry[] = [];
                 for (const owner of collection.getOwners()) {
-                    const marker = collection.getMarkers(owner);
-                    if (marker) {
```
Comment (marechal-p) [inline packages/markers/src/browser/marker-manager.ts]: why did you remove this check ?

**Comment by simark** [comment]
@lmcbout I wrote a little test that checks that we don't save URIs with no markers:

https://github.com/simark/theia/commit/8d71112f3cfd99ae411752d7e003b2aca1f69bd3 

It can probably be cleaned up, but it seems to work.

File: packages/markers/src/browser/marker-manager.spec.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        // We want to be able to wait until the marker manager has set the data in the local storage.
+        let setDataPromise = new Deferred<void>();
+        const storageService = container.get<MockStorageService>(MockStorageService);
+        storageService.onSetData(() => setDataPromise.resolve());
+
+        // Provide one marker
+        const mm = container.get(MarkerManager);
+        const uri = new URI('/path/to/file1.ts');
+        const owner = 'test-owner';
+        mm.setMarkers(uri, 'test-owner', ['allo']);
```
Comment (akosyakov) [inline packages/markers/src/browser/marker-manager.spec.ts]: is not `await mm.serMarkers` should be enough to make sure that markers are stored? `setDataPromise.promise` seems to be redundant

File: packages/core/src/browser/test/mock-storage-service.ts [hunk 3]
```diff
@@ -14,14 +14,24 @@ import { injectable } from 'inversify';
 @injectable()
 export class MockStorageService implements StorageService {
     readonly data = new Map<string, {} | undefined>();
+    onSetDataCallback?: (key: string, data?: any) => void;
```
Comment (akosyakov) [inline packages/core/src/browser/test/mock-storage-service.ts]: We should use event for such APIs to support multiple listeners. But in this case an event seems to be redundant: https://github.com/theia-ide/theia/pull/2093/files/c116ae8ac34a36fec80c66a734982661ed3ff3cb#r196008736

File: packages/markers/src/browser/marker-manager.spec.ts (hunk 2 above)
Comment (simark) [reply packages/markers/src/browser/marker-manager.spec.ts]: No, if you follow the code path inside setMarkers, you'll see that it doesn't wait that the markers are written to the local storage.

Comment (akosyakov) [reply packages/markers/src/browser/marker-manager.spec.ts]: i think it does:
- setMarkers
- internalSetMarkers
- fireOnDidChangeMarkers
- saveMarkersToStorage

Comment (simark) [reply packages/markers/src/browser/marker-manager.spec.ts]: But we don't await that fireOnDidChangeMarkers callbacks have finished.  So we can return from setMarkers before the data is actually saved to the local storage.

Comment (akosyakov) [reply packages/markers/src/browser/marker-manager.spec.ts]: `fireOnDidChangeMarkers` is synchronous

but ok, i've just realized that `onSetDataCallback` is added to a mock not to a real implementation

File: packages/markers/src/browser/marker-manager.ts [hunk 4]
```diff
@@ -39,13 +39,12 @@ export class MarkerCollection<T> {
     }
 
     setMarkers(owner: string, markerData: T[]): Marker<T>[] {
-        const before = this.owner2Markers.get(owner);
```
Comment (akosyakov) [inline packages/markers/src/browser/marker-manager.ts]: Could you elaborate why the contract should be changed? It used to return old markers.

File: packages/markers/src/browser/marker-manager.spec.ts (hunk 2 above)
Comment (simark) [reply packages/markers/src/browser/marker-manager.spec.ts]: > fireOnDidChangeMarkers is synchronous

The callback is invoked synchronously, but it can start some asynchronous operations (such as writing to local storage), and we don't await that.

File: packages/markers/src/browser/marker-manager.ts (hunk 4 above)
Comment (marechal-p) [reply packages/markers/src/browser/marker-manager.ts]: The function seems to be the same, `before` wasn't used except in the `return`, which was changed:
```ts
return before || []; // before
return this.owner2Markers.get(owner) || []; // after
```

Comment (simark) [reply packages/markers/src/browser/marker-manager.ts]: But it used to return the old markers, and now it will return the new markers.

Comment (marechal-p) [reply packages/markers/src/browser/marker-manager.ts]: Oh right m bad

**Comment by lmcbout** [comment]
@marcdumais-work changed the commit message

File: packages/markers/src/browser/marker-manager.ts (hunk 4 above)
Comment (simark) [reply packages/markers/src/browser/marker-manager.ts]: @marechal-p I think it's preferrable not to remove your message, it just makes the discussion more difficult to read.

**Comment by marcdumais-work** [comment]
> @marcdumais-work changed the commit message

I suggest something like the following commit message and supplementary text:

```
[marker-manager] clear files without markers

marker-manager manages markers, for problems found by e.g. language servers.
Each time the user opens a file in an editor, the file is analysed by the language 
server, and markers are created for any problem found. If the file has no problems,
an empty list is sent to the manager, allowing it to clear markers previously 
associated to that URI.

ATM the manager keeps an internal entry for files with no associated markers. There 
entries are preserved in local storage and restored at client startup. After a while, 
there can be many. Since they serve no apparent useful purpose, this commit clears 
those entries. By consequence,they are no longer persisted in local storage.

```

**Comment by lmcbout** [comment]
@marcdumais-work Commit message modify as requested.

**Comment by simark** [comment]
> ATM the manager keeps an internal entry for files with no associated markers. There entries are preserved in local storage and restored at client startup. After a while,

"There" -> "These".  Otherwise, the text LGTM.

**Comment by lmcbout** [comment]
"There" -> "These" fixed

File: packages/markers/src/browser/marker-manager.ts (hunk 4 above)
Comment (akosyakov) [reply packages/markers/src/browser/marker-manager.ts]: so it should be reverted since it changes the semantic for no reason?

Comment (simark) [reply packages/markers/src/browser/marker-manager.ts]: > so it should be reverted since it changes the semantic for no reason?

Yes.

Comment (lmcbout) [reply packages/markers/src/browser/marker-manager.ts]: Keeping the old markers does not make sence when you have a new set of markers for this file. When you make a modification to remove the problems, you don't want to have a list of obsolete markers being maintained, you want the latest information about this file. So with the change, wether you have a new set of markers or no more markers, you want to update the list.

Comment (simark) [reply packages/markers/src/browser/marker-manager.ts]: The list of old markers that is returned is not kept anywhere, it's just returned to the caller, if they want to do something with this information.  But in the marker manager, we always keep the latest info, regardless.

Actually, no caller seem to use the return value anyway, so maybe we could just remove the return value?

Comment (marcdumais-work) [reply packages/markers/src/browser/marker-manager.ts]: > Actually, no caller seem to use the return value anyway, so maybe we could just remove the return value?

True, as far as we know. But an "outside" extender might be using the marker-manager and depending on the old behavior. So I think the default approach of having to reasonably justify changing a public API, is proper.

Comment (marechal-p) [reply packages/markers/src/browser/marker-manager.ts]: The return value is indirectly used in a test (`problem-manager.spec.ts`), and it is only by luck that the test still passes: The test check the number of errors before / after, turns out it is 2 for both cases.

Anyway this modification should be reverted as it adds nothing. If we want the new markers we call `collection.getMarkers()` after the `collection.setMarkers()` (especially in `MarkerManager.internalSetMarkers`).

Comment (lmcbout) [reply packages/markers/src/browser/marker-manager.ts]: I put back the line in setMarkers() to return the old list of markers.
Question: On a set() method we should have two choices: either no return value since we are setting, adding,  removing a value  or we should return the result of the setting, not returning what was available before we call set() method. When a user is looking of what was available before a set() method, he can perform a get() and use that data as he wishes. We should not depend of the return of a set() to GET the OLD data. I think this is a bad use of a set() method

Comment (simark) [reply packages/markers/src/browser/marker-manager.ts]: > The return value is indirectly used in a test (problem-manager.spec.ts), and it is only by luck that the test still passes: The test check the number of errors before / after, turns out it is 2 for both cases.

Ahh you're right I missed that.  Well, maybe this test should be improved then :)

> Anyway this modification should be reverted as it adds nothing. If we want the new markers we call collection.getMarkers() after the collection.setMarkers() (especially in MarkerManager.internalSetMarkers).

Why would we want the new markers in internalSetMarkers, isn't it supposed to return the old ones?

File: packages/markers/src/browser/marker-manager.ts [hunk 5]
```diff
@@ -183,9 +182,13 @@ export abstract class MarkerManager<D extends object> {
 
     protected internalSetMarkers(uri: URI, owner: string, data: D[]): Marker<D>[] {
         const collection = this.getCollection(uri);
-        const result = collection.setMarkers(owner, data);
+        const old = collection.setMarkers(owner, data);
+        const result = collection.getMarkers(owner);
```
Comment (marechal-p) [inline packages/markers/src/browser/marker-manager.ts]: `result` should be named `current` or something similar.

File: packages/markers/src/browser/marker-manager.ts (hunk 4 above)
Comment (marechal-p) [reply packages/markers/src/browser/marker-manager.ts]: > Why would we want the new markers in internalSetMarkers, isn't it supposed to return the old ones?

Yes but I think @lmcbout added a check to delete an entry as [part of an optimization I think](https://github.com/theia-ide/theia/pull/2093/files#diff-fa0fb0be865e7c3cf2af196391b00cf6R187).

File: packages/markers/src/browser/marker-manager.ts (hunk 5 above)
Comment (marcdumais-work) [reply packages/markers/src/browser/marker-manager.ts]: Could we get rid of this line and instead use ```if (data.length === 0)``` to find out if the marker list is empty? This would avoid the extra call to get the updated markers list.

**Comment by marcdumais-work** [comment]
@simark so you have pending comments for this PR? It seems your current "-1" was related to a missing test, that you since supplied.

@akosyakov are you ok with the current version, where the original API/contract for ```marker-manager#setMarkers()``` is restored?

File: packages/markers/src/browser/marker-manager.ts [hunk 6]
```diff
@@ -148,18 +148,17 @@ export abstract class MarkerManager<D extends object> {
...
+                    ownerEntries.push({
+                        owner,
+                        markerData: Array.from(markers.map(m => m.data))
+                    });
                 }
                 result.push({
                     uri,
                     markers: ownerEntries
                 });
+
```
Comment (simark) [inline packages/markers/src/browser/marker-manager.ts]: Unnecessary line change here.

**Comment by simark** [comment]
> @simark so you have pending comments for this PR? It seems your current "-1" was related to a missing test, that you since supplied.

Yes, fine with me.
